src/Interpreters/ITokenExtractor.h

```diff
@@ -160,48 +160,53 @@ struct NgramTokenExtractor final : public ITokenExtractorHelper<NgramTokenExtrac
-
     bool nextInStringLike(const char * data, size_t length, size_t * pos, String & token) const override;
 
     size_t getN() const { return n; }
 
 private:
-
     size_t n;
 };
 
-/// Parser extracting tokens (sequences of numbers and ascii letters).
+/// Parser extracting tokens which consist of alphanumeric ASCII characters or Unicode characters (not necessarily alphanumeric)
 struct SplitTokenExtractor final : public ITokenExtractorHelper<SplitTokenExtractor>
 {
     static const char * getName() { return "tokenbf_v1"; }
     static const char * getExternalName() { return "default"; }
 
-    std::vector<String> getTokens(const char * data, size_t length) const override;
-
     bool nextInString(const char * data, size_t length, size_t * __restrict pos, size_t * __restrict token_start, size_t * __restrict token_length) const override;
-
     bool nextInStringPadded(const char * data, size_t length, size_t * __restrict pos, size_t * __restrict token_start, size_t * __restrict token_length) const override;
-
     bool nextInStringLike(const char * data, size_t length, size_t * __restrict pos, String & token) const override;
-
     void substringToBloomFilter(const char * data, size_t length, BloomFilter & bloom_filter, bool is_prefix, bool is_suffix) const override;
-
     void substringToGinFilter(const char * data, size_t length, GinFilter & gin_filter, bool is_prefix, bool is_suffix) const override;
 };
 
+/// Parser extracting tokens which are separated by certain strings.
+/// Allows to emulate e.g. BigQuery's LOG_ANALYZER.
+struct StringTokenExtractor final : public ITokenExtractorHelper<StringTokenExtractor>
+{
+    explicit StringTokenExtractor(const std::vector<String> & separators_);
+
+    static const char * getName() { return "string"; }
+    static const char * getExternalName() { return getName(); }
+
+    bool nextInString(const char * data, size_t length, size_t * pos, size_t * token_start, size_t * token_length) const override;
+    bool nextInStringLike(const char * data, size_t length, size_t * pos, String & token) const override;
+
+private:
+    std::vector<String> separators;
+};
+
 /// Parser doing "no operation". Returns the entire input as a single token.
 struct NoOpTokenExtractor final : public ITokenExtractorHelper<NoOpTokenExtractor>
 {
-    static const char * getName() { return "noop"; }
+    static const char * getName() { return "no_op"; }
     static const char * getExternalName() { return getName(); }
 
     std::vector<String> getTokens(const char * data, size_t length) const override;
-
-    bool nextInString(const char * data, size_t length, size_t * __restrict pos, size_t * __restrict token_start, size_t * __restrict token_length) const override;
-
-    bool nextInStringLike(const char * data, size_t length, size_t * __restrict pos, String & token) const override;
+    bool nextInString(const char * data, size_t length, size_t * pos, size_t * token_start, size_t * token_length) const override;
```

Comment:
The functions are inconsistent across the base class/child classes regarding `__restrict`. Again, I am not sure about the implications, but I think it is better to keep them aligned.

Reply:
`__restrict` is a slightly obscure optimization hint (it tells the compiler that pointers passed as arguments don't point to the same memory). Chances are that nobody measured the difference and that it doesn't matter.

The no-op tokenizer does nothing, it is certainly not performance-critical, so lets remove `__restrict` instead of cargo-culting it everywhere.